documentation/python.py

```diff
@@ -2070,47 +2078,83 @@ def render_class(state: State, path, class_, env):
     for hook in state.hooks_post_scope:
         hook(type=EntryType.CLASS, path=path)
 
-# Extracts image paths and transforms them to just the filenames
-class ExtractImages(Transform):
-    # Max Docutils priority is 990, be sure that this is applied at the very
-    # last
-    default_priority = 991
 
-    # There is no simple way to have stateful transforms (the publisher always
-    # gets just the class, not the instance) so we have to make all data
-    # awfully global. UGH.
-    # TODO: maybe the pending nodes could solve this?
-    _url_formatter = None
-    _external_data = set()
+# A helper function to work-around for stateless Docutils transforms
+def get_image_extracter(source, external_data:set, config):
+
+    # Extracts image paths and transforms them to just the filenames
+    class ExtractImages(Transform):
+        # Max Docutils priority is 990, be sure that this is applied at the very
+        # last
+        default_priority = 991
 
-    def __init__(self, document, startnode):
-        Transform.__init__(self, document, startnode=startnode)
+        def __init__(self, document, startnode):
+            Transform.__init__(self, document, startnode=startnode)
 
-    def apply(self):
-        ExtractImages._external_data = set()
-        for image in self.document.traverse(docutils.nodes.image):
-            # Skip absolute URLs
-            if urllib.parse.urlparse(image['uri']).netloc: continue
+        def apply(self):
+            input_path = os.path.realpath(config['INPUT'])
+            for image in self.document.traverse(docutils.nodes.image):
+                # Skip absolute URLs
+                if urllib.parse.urlparse(image['uri']).netloc:
+                    continue
 
-            # TODO: is there a non-private access to current document source
-            # path?
-            absolute_uri = os.path.join(os.path.dirname(self.document.settings._source), image['uri']) if isinstance(self.document.settings._source, str) else image['uri']
-            ExtractImages._external_data.add(absolute_uri)
+                if isinstance(source, str):
+                    image_abs_path = os.path.realpath(os.path.join(os.path.dirname(source), image['uri']))
+                    # Use relative path if image within INPUT directory, otherwise absolute
+                    if image_abs_path.startswith(input_path):
+                        image_path = os.path.relpath(image_abs_path, input_path)
```

Comment:
This always returns relative path to INPUT directory, will be patched later by `PatchReferences`